pageserver/src/tenant.rs

```diff
@@ -1389,13 +1445,107 @@ impl Tenant {
             timeline
         };
 
+        // multiple concurrent requests could had "transitioned" a timeline to stopping state,
+        // because repeat transitions to same state are allowed. because the delete_timeline will
+        // need to synchronize and cannot be original semi-lockless algorithm for "upload
+        // indexpart" part, we create a single task to delete the timeline.
+
+        let mut rx = {
+            let mut g = timeline.delete_self.lock().await;
+
+            // see if we should join any existing attempt
+            let maybe_rx = if let Some(maybe_done) = g.as_ref() {
+                use timeline::MaybeDone;
+
+                match maybe_done {
+                    MaybeDone::Done(Ok(())) => return Ok(()),
+                    MaybeDone::Done(Err(e)) if e.is_permanent() => {
+                        return Err(DeleteTimelineError::from(e))
+                    }
+                    MaybeDone::Pending(rx) => {
+                        // important: upgrading and later resubscription happens while holding the lock
+                        rx.upgrade()
+                    }
+                    MaybeDone::Done(Err(_retryable)) => None,
+                }
+            } else {
+                None
+            };
+
+            if let Some(rx) = maybe_rx {
+                rx
+            } else {
+                // try another time
+                let (tx, rx) = tokio::sync::broadcast::channel(1);
+                let rx = Arc::new(rx);
+                let this = self.clone();
+                let timeline = timeline.clone();
+
+                // TODO: this could be tenant scoped task_mgr task?
+                tokio::spawn({
```

Comment:
Yes it should be.
Otherwise, we'll not wait for this task, e.g., during pageserver shutdown / tenant detach.

Reply:
Done, at quite more horrible internal impl in #4194.